server/mqtt.go

```diff
@@ -3440,15 +3812,73 @@ func (c *client) mqttProcessPubAck(pi uint16) {
 		if len(sess.pending) == 0 {
 			sess.ppi = 0
 		}
-		ackSubject = ack.subject
+		jsAckSubject = ack.jsAckSubject
+	}
+	sess.mu.Unlock()
+
+	// Send the ack to JS to remove the pending message from the consumer.
+	sess.jsa.sendAck(jsAckSubject)
+}
+
+func (s *mqttSession) pubRelSubject() string {
+	return mqttQOS2PubRelStoredSubjectPrefix + s.idHash
+}
+
+func (s *mqttSession) pubRelDeliverySubject() string {
+	return mqttQOS2PubRelDeliverySubjectPrefix + s.idHash
+}
+
+// Process a PUBREC packet (QoS2, acting as Sender).
+//
+// Runs from the client's readLoop.
+// No lock held on entry.
+func (c *client) mqttProcessPubRec(pi uint16) error {
+	sess := c.mqtt.sess
+	if sess == nil {
+		return nil
+	}
+
+	// Initiate PUBREL delivery
+	natsMsg, headerLen := mqttNewDeliverablePubRel(pi)
+	_, err := sess.jsa.storeMsg(sess.pubRelSubject(), headerLen, natsMsg)
+	if err != nil {
+		return err
+	}
+
+	sess.mu.Lock()
+	if sess.c != c {
+		return nil
+	}
+
+	ack, ok := sess.pending[pi]
+	if !ok {
+		sess.mu.Unlock()
+		return nil
 	}
-	// Send the ack if applicable.
-	if ackSubject != _EMPTY_ {
-		// We pass -1 for the hdr so that the send loop does not need to
-		// add the "client info" header. This is not a JS API request per se.
-		sess.jsa.sendq.push(&mqttJSPubMsg{subj: ackSubject, hdr: -1})
+	ackSubject := ack.jsAckSubject
+
+	// Remove the old consumer/seq->PI mapping.
+	jsDur := ack.jsDur
+	if sseqToPi, ok := sess.cpending[jsDur]; ok {
+		delete(sseqToPi, ack.sseq)
+	}
+	if len(sess.pending) == 0 {
+		sess.ppi = 0
 	}
+
+	// Move the PI tracking to the PUBREL consumer. It will not have a
+	// sequence nor an ACK subject until we get a PUBREL callback on the
+	// subscription. See <>/<>
+	sess.pending[pi].jsDur = sess.pubRelConsumer.Durable
+	sess.pending[pi].jsAckSubject = ""
+	sess.pending[pi].sseq = 0
```

Comment:
Not sure if the compiler optimizes that, but anyway, you already have `ack, ok := sess.pending[pi]` above, so use `ack.jsDur ...` instead here.

Reply:
I am confused. 

Line 3872 changes the consumer name for this PI from "messages" (which is what it was before, and what's in the ack) to "pubrel", where I do not yet have the actual ack subject; that would be in `mqttDeliverPubRelCb` where I get it.

Reply:
What I meant is that all those lines with `sess.pending[pi]...=` mean that there is a map lookup, then assignment to the returned value by the lookup. The lookup was done earlier when you had `ack, ok := sess.pending[pi]`. I don't see anywhere that you are changing `pi`, so those calls are actually modifying the structure that `ack` points to. If that is not what you intended, then the code needs to be changed.